Documented the `RestartableJenkinsRule` (#7754)
jenkins-infra/jenkins.io: content/doc/developer/testing/index.adoc

* link:https://github.com/jenkins-infra/pipeline-library#runbenchmarks[Running benchmark through Jenkinsfile]

== Further Pipeline Testing

> Sometimes you might need to test more sophisticated things, as a restarting Jenkins and want to verify that the changes are still present.

Reviewer: ```suggestion
Sometimes you might need to test more sophisticated things, as restarting Jenkins and want to verify that the changes are still present.
```

=== Testing Durable Pipeline Steps

> If you want to ensure that your plugin behaves correctly after a restart, there are methods helping you.

Reviewer: ```suggestion
If you want to ensure that your plugin behaves correctly after a restart, there are methods to help you.
```

> `RestartableJenkinsRule` is part of the Jenkins testing framework and provides an easy way to simulate restarting Jenkins in your tests without actually restarting the Jenkins instance.

Reviewer: ```suggestion
`RestartableJenkinsRule` is part of the Jenkins testing framework and provides an easy way to simulate restarting Jenkins in your tests without actually having to restart the Jenkins instance.
```

> You can use the `then()` method to run one Jenkins session and shut down.

Reviewer: ```suggestion
You can use the `then()` method to run one Jenkins session and then shut down.
```

> The following example shows how it can be used with a `FreeStyleJob` run which is verified after a Jenkins restart:

Reviewer 2: I lack fundamental skills in the use of `RestartableJenkinsRule`, so I'm not a great reviewer of this pull request.  My apologies in advance for mistakes that I may make in the review process.

Since the example uses a `FreeStyleProject`, I think that the section heading should be changed from "Testing Durable Pipeline Steps" to "Testing Jenkins Restart".  The example looks reasonable to me as a way to show that a `FreeStyleProject` build status persisted across a restart.

I think the job type used in the test should be referenced using the phrase "Freestyle job", the phrase "Freestyle project", or the Java class name "`FreeStyleProject`".  The first two forms are used elsewhere in the www.jenkins.io documentation and the third is used in the example source code.
```suggestion
The following example shows how it can be used with a `FreeStyleProject` build which is verified after a Jenkins restart:
```

[source,java]
----
import hudson.model.FreeStyleProject;
import org.jvnet.hudson.test.RestartableJenkinsRule;
import org.junit.Rule;
import org.junit.Test;

public class MyPluginTest {

    @Rule
    public RestartableJenkinsRule rr = new RestartableJenkinsRule();

    @Test
    public void testPluginAfterRestart() {
        rr.then(r -> {
            // Set up the test environment
            // e.g., create a job, configure your plugin, etc.
            FreeStyleProject project = r.createFreeStyleProject("myJob");
            // let's execute one build
            project.scheduleBuild2(0).get();
        });

        // now we want to verify things after the restart
        rr.then(r -> {
            // get first project
            FreeStyleProject p = (FreeStyleProject) r.getInstance().getAllItems().get(0);
            // verify that it is still successfully
            r.assertBuildStatusSuccess(p.getBuild("1"));
        });
    }
}
----

> For more detailed information you can take a look at the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/RestartableJenkinsRule.html[`RestartableJenkinsRule` JavaDoc].

Reviewer: ```suggestion
For more detailed information, you can take a look at the link:https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/RestartableJenkinsRule.html[`RestartableJenkinsRule` JavaDoc].
```
